### ISSUE TITLE
Don't fire update if the top-level state doesn't change after a reducer

### ISSUE THREAD

**Issue by LPGhatguy** [Roblox/rodux#32]
This probably doesn't come up often in practice, but it seems like a fairly low-effort optimization for this case.

**Comment by OverHash** [comment]
By top-level state are you referring to not checking recursively? That would make more sense as checking recursively can be quite costly.

Part of this issue is optimization, but it heavily depends on what's happening inside the `.changed`, for instance, if users manually check to see if the value they want has actually changed, this issue doesn't affect them (and would actually slow their code down), but otherwise, it could make a slight performance gain.

**Comment by ZoteTheMighty** [comment]
If I understand correctly, the only time that this will be an improvement is if we're using the store and thunks to to trigger a side effect without actually changing any state, right? That seems too infrequent to be a concern, especially since lots of thunks will also dispatch more changes that _do_ change the state before the next flush anyways.

**Comment by edward827272-cmyk** [comment]
I agree that this sounds like a relatively narrow optimization. If most thunks dispatch additional actions that eventually modify state before the next flush, the practical benefit may be limited. The improvement seems more noticeable for side effects that intentionally avoid state updates altogether. Unless that pattern is common in your application, I’d probably prioritize other performance concerns first. Tools like [onetruco ipc64](https://onetruco.org/) may also offer useful context.
